### PR TITLE
Generate documentation using dts bindings

### LINKED ISSUE #28865: Doc: Generate documentation using dts bindings
**Is your enhancement proposal related to a problem? Please describe.**
dts bindings provide useful details on:
- what is implemented
- how it is implemented
- how it could/should be configured

Similarly to Kconfig and API it would be quite interesting to generate documentation out of dts/bindings.

**Describe the solution you'd like**
A doc section (ending up being available on https://docs.zephyrproject.org/) that would provide all details on available bindings.

**Describe alternatives you've considered**

**Additional context**
Example where bindings could be useful:
Looking to drivers/sensors  we can't find driver for lis2dh12. Fact is that lis2dh12 is supported by lis2dh driver, which is a piece of information hard to find out.
Looking at the bindings, we can see straight away that `st,lis2dh12-i2c.yaml` is available and hence supported.

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
This pull request extends the devicetree reference documentation to include information on each available binding.

Fixes: #28865

### CODE REVIEW

**Comment by mbolivar-nordic** [comment]
In version 1 as first pushed, the bindings index page looks like this:

bindings-index

Each individual binding page looks like this:

binding-example

TODOs:

- generate child binding documentation
- extract properties inherited via base.yaml from properties that are specific to just some nodes

**Comment by erwango** [comment]
I find this really nice. The index page looks great, this is dense but clear and readable.
The individual property page is nice as is, but maybe something more dense would be nice as well.
Anyway, this will be an incentive to provide more effort in the bindings descriptions.

**Comment by avisconti** [comment]
The page looks really nice. Though I still wonder if the original purpose, which is to make clear to the final user whether a particular device is supported or not, has been met.

It is pretty clear to me how to check this info out from this page. Not sure if it is for any user.
Does it make sense my doubt?

**Comment by trond-snekvik** [comment]
Looks fantastic, a massive improvement for the DeviceTree workflow.

Some suggestions:
- If the property description is the main content of each section, instead of this conditional variant, it'd look more like the API documentation and encourage longer descriptions. It'd also keep the "Required" and type info at the top. IMO, we should permit rst formatting in the description instead of making it a code block, but it might yield some bad side effects.
- A personal preference perhaps, but IMO the property entries would look better with more complex formatting instead of putting everything in a list or table with a field prefix. For instance:
  - Make the type monospaced and place it right below the entry name. 
  - "Required/Optional" instead of "required: True/False". In italics, perhaps?
  - A list of monospaced entries for the enum
- If each property type described in the bindings template file had their own subsection in the Devicetree guide, the generated type field could link to those sections
- Very few nodes have more than one or two specifier cell listings, they could probably follow the same subsection+table listing as the properties. It'd also be neat if the cell specifier bindings format supported descriptions for the specifier cells, so we could get more information, like what flags or macros are supported for each cell. This is probably something for a separate PR though.
- Include list of included bindings in the binding page, with links? Could also mark properties as "inherited from XYZ"
- Include vendor name in binding page

**Comment by erwango** [comment]
@avisconti well I think that at one point in the index page, we'll find a list of available bindings for, say, sensors. If a reference is there: it is supported, if it's not there ... Maybe this could be made even more clear in the index page.
But then, to make it clearer to random users. I think it will be then up to developers to properly document the bindings.
For instance in dts/bindings/sensor/st,iis2mdc-i2c.yaml, description could be elaborated with more details
```
description: |
    STMicroelectronics IIS2MDC magnetometer accessed through I2C bus. 
    Support of IIS2MDC on a I2C1 bus node is done as follows

   &i2c1 {
	iis2mdc@1e {
		compatible = "st,iis2mdc";
		reg = <0x1e>;
		drdy-gpios =  <&gpioa 2 GPIO_ACTIVE_HIGH>; /* A2 */
		label = "IIS2MDC";
	};    
```
And this description will directly appear in the generated individual binding page, just before the clickable property list.
You can even throw more details there about behavior available. 

Maybe even pointing to related Kconfig symbols would be doable.
Am I right @mbolivar-nordic ?

**Comment by avisconti** [comment]
@erwango 
OK, looks clear to me now.

**Comment by mbolivar-nordic** [comment]
@avisconti:

> Though I still wonder if the original purpose, which is to make clear to the final user whether a particular device is supported or not, has been met.

Definitely not completely, I would say. This is just one step.

What @erwango says about documenting how to configure individual nodes better in the bindings descriptions is a valid point and I think maintainers should start to do this after the PR is merged.

And indeed this is a great hint that Zephyr has some driver somewhere that can create a `struct device` for this node.

But some big questions would still be unanswered by existing documentation:

- what device driver (or drivers) can handle this compatible?
- what API does it implement?
- how do I enable support for this driver in Kconfig, and what other configuration do I need to do?
- how do I use this API?

Though we have done a great job at adding Doxygen comments for our existing APIs, I think our Sphinx documentation for device drivers still has big gaps.

Can you explain to me how to use spi.h from just this page, for example?
https://docs.zephyrproject.org/latest/reference/peripherals/spi.html#spi-api

And even assuming good API docs, the gap from "this is the API" to "these are the drivers that implement this API, this is how to enable them in Kconfig, and this is how to configure the instances in devicetree" is still very large.

My goal with this PR is to let SoC maintainers start to close the "this is how to configure the instances in devicetree" part of the gap. I don't have a proposal for how to handle the rest. Right now we have to spread the tribal knowledge about the source tree structure and how to grep for DT_DRV_COMPAT, etc., but it's not good enough IMO.

**Comment by mbolivar-nordic** [comment]
I've reworked the way the individual bindings pages look. Here is a representative example.

v2-binding-page

**Comment by avisconti** [comment]
> My goal with this PR is to let SoC maintainers start to close the "this is how to configure the instances in devicetree" part of the gap. I don't have a proposal for how to handle the rest. Right now we have to spread the tribal knowledge about the source tree structure and how to grep for DT_DRV_COMPAT, etc., but it's not good enough IMO.

@mbolivar-nordic 
Yes, sure. I think that this is a big improvement from that perspective and it is a task now for the maintainers to proceed in a better and hopefully exhaustive explanation in the binding description as suggested by @erwango.

For what I understood the primary path for a user to see whether a sensor (for example) is supported or not is to look inside `zephyr/drivers/sensor` and see the sensor names. I think that looking at bindings may work as well, but the documentation should be quite visible and should be somehow taken as preferred path to see what devices are supported or not.

But one step at a time. Thanks for your huge effort on this part!

**Comment by mbolivar-nordic** [comment]
@trond-snekvik thanks for the detailed comments. Here are my responses:


> IMO, we should permit rst formatting in the description instead of making it a code block, but it might yield some bad side effects.

I agree and originally tried it this way. As you guessed, it caused problems, because a lot of the binding descriptions aren't valid RST.

> A personal preference perhaps, but IMO the property entries would look better with more complex formatting instead of putting everything in a list or table with a field prefix.

I reformatted the property descriptions as a table with more complex formatting:

- types are monospaced, (required) is after the property name if true
- Enum and const are combined into a single "legal values" column. I did not monospace this, though.

> If the property description is the main content of each section, instead of this conditional variant, it'd look more like the API documentation and encourage longer descriptions.

I have done it this way for the description of the entire binding. I really agree with you about it being good encouragement to add more information there.

I do think some individual properties legitimately have one-line descriptions, though, and prefer to permit the compact representation there when possible. 

> Include vendor name in binding page

Done, with back-link to index.

> Include list of included bindings in the binding page, with links? Could also mark properties as "inherited from XYZ"

I want this too, but edtlib doesn't support this. I've separated base
bindings from node-specific bindings as an 85% solution for now.

> If each property type described in the bindings template file had their own subsection in the Devicetree guide, the generated type field could link to those sections

Agree, that would be great, but that's a PR for another day; this PR
is just about getting generated documentation for bindings.

> Very few nodes have more than one or two specifier cell listings, they could probably follow the same subsection+table listing as the properties. It'd also be neat if the cell specifier bindings format supported descriptions for the specifier cells, so we could get more information, like what flags or macros are supported for each cell. This is probably something for a separate PR though.

Agree, that would also be great, but I think I want to let individual
binding authors handle this in the descriptions of each node for now.
Specifier cells are a whole deal to document.

There's a lot of follow-up work I've sketched out in a previous
comment, but I hope that what's here is a "good enough" starting
point.

**Comment by mbolivar-nordic** [comment]
> For what I understood the primary path for a user to see whether a sensor (for example) is supported or not is to look inside `zephyr/drivers/sensor` and see the sensor names.

That doesn't work for everyone; in particular it doesn't work if you have out of tree drivers, which Nordic does in our Zephyr-based SDK. Centralized documentation would be much better for discoverability.

I also personally think that looking in drivers/sensor and checking source file names is not as beginner friendly as it could be.

> I think that looking at bindings may work as well, but the documentation should be quite visible and should be somehow taken as preferred path to see what devices are supported or not.

I think of bindings docs as just a way to document how to configure individual instances.



> But one step at a time. Thanks for your huge effort on this part!

Thanks, and thanks for review!

**Comment by galak** [comment]
@mbolivar-nordic like the new table style.

* feel like `required` is getting lost. Wonder if maybe using a different color for the row as well? or bold the name + `(required)`?
* Mixed of fixed and non-fixed font looks weird.  Maybe also make the `Description` column fixed.
* Maybe also make the prop names fixed.

**Comment by avisconti** [comment]
> > For what I understood the primary path for a user to see whether a sensor (for example) is supported or not is to look inside `zephyr/drivers/sensor` and see the sensor names.
> 
> That doesn't work for everyone; in particular it doesn't work if you have out of tree drivers, which Nordic does in our Zephyr-based SDK. Centralized documentation would be much better for discoverability.

YEs, don't get me wrong. The documentaion is obviously the principal starting point if you want to be serious. What I just wanted to highlight the fact that some user appears to be lazy enough to use the code as their searching point. At least this is my experience. THat-s where my initial concern was coming from.

**Comment by mbolivar-nordic** [comment]
Here is an updated image of a properties table (taken from `ti,lp503x`):

Screenshot from 2020-10-27 21-09-14




@galak:

> * feel like `required` is getting lost. Wonder if maybe using a different color for the row as well? or bold the name + `(required)`?

I tried a few things but I ran into horizontal scrolling problems when I added too many columns to these tables, so I've done the above instead.

> 
> * Mixed of fixed and non-fixed font looks weird.  Maybe also make the `Description` column fixed.

I've made the description text itself always fixed in the "Details" column. Seem OK?

> * Maybe also make the prop names fixed.

Check out the latest picture the `label` property name is a link now, that points to the doc/guides/dts/intro.rst content describing that property. Similar changes were made for important things like `reg` and `interrupts`.

I couldn't figure out a way to make link text fixed, and I think it would be worse to make them sometimes-fixed (if no link) and sometimes-not-fixed (if link), so I hope it's OK the prop names are not fixed. I think having a reference to (hopefully useful) docs on important properties is worth it.

**Comment by mbolivar-nordic** [comment]
I took this out of draft, hopefully it's OK now.

**Comment by galak** [comment]
I think we can remove `Description:` out of the `details` column.

**Comment by galak** [comment]
I think having the `name` as a fixed font makes sense since that name should exist in the actual dts.

**Comment by mbolivar-nordic** [comment]
> I think we can remove `Description:` out of the `details` column.

Even if it means horizontal scrolling, e.g. on laptop screens? People *hate* horizontal scrolling. I know I do.

> I think having the name as a fixed font makes sense since that name should exist in the actual dts.

Is that really more important than links to documentation?

**Comment by mbolivar-nordic** [comment]
> > I think we can remove `Description:` out of the `details` column.
> 
> Even if it means horizontal scrolling, e.g. on laptop screens? People _hate_ horizontal scrolling. I know I do.

Oh, you mean removing just the text "description" from the column, not moving the description text to a new column. OK :+1:

**Review by nashif — APPROVED** [review]
great stuff

File: doc/scripts/gen_devicetree_rest.py
Comment (utzig) [inline doc/scripts/gen_devicetree_rest.py:353]: Shouldn't `os.path.join` be used here? I am sure eventually someone will want to build the documentation on Windows!

File: doc/scripts/gen_devicetree_rest.py
Comment (utzig) [inline doc/scripts/gen_devicetree_rest.py:431]: `os.path.join`?

File: doc/scripts/gen_devicetree_rest.py
Comment (utzig) [inline doc/scripts/gen_devicetree_rest.py:273]: `os.path.join`?

File: doc/scripts/gen_devicetree_rest.py
Comment (utzig) [inline doc/scripts/gen_devicetree_rest.py:268]: `os.path.join`?

File: doc/scripts/gen_devicetree_rest.py
Comment (utzig) [inline doc/scripts/gen_devicetree_rest.py:243]: `os.path.join`?

**Review by utzig — CHANGES_REQUESTED** [review]
Works well and nicely on my machine, I am mostly concerned with building documentation on Windows, so I added some repetitive comments about using `os.path.join`!

**Review by utzig — APPROVED** [review]
Duh, it's all pathlib, disregard my previous comments. LGTM!

**Comment by mbolivar-nordic** [comment]
@galak I've tried to come up with a compromise that preserves links for important properties but also renders every property name in fixed width in the Name column:

Screenshot from 2020-10-30 09-26-28

**Review by MaureenHelm — APPROVED** [review]
I haven't studied the implementation details, but I really like the output. Nice work @mbolivar-nordic !

**Review by pabigot — APPROVED** [review]
This did come out nice; thank you.

File: dts/bindings/vendor-prefixes.txt
Comment (mbolivar-nordic) [reply dts/bindings/vendor-prefixes.txt:4]: >  We should have a script that converts [vendor-prefixes.yaml](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/Documentation/devicetree/bindings/vendor-prefixes.yaml) to the old format that we're still using.

Yeah, I had the same thought, but figured one thing at a time.

**Pull request merged**
